libs/remote_storage/src/s3_bucket.rs

```diff
@@ -37,82 +38,249 @@ use crate::{
+            self.get(kind).observe(started_at.elapsed().as_secs_f64())
+        }
+    }
+
+    pub(super) struct PassFailCancelledRequestTyped<C> {
+        success: RequestTyped<C>,
+        fail: RequestTyped<C>,
+        cancelled: RequestTyped<C>,
+    }
+
+    #[derive(Debug, Clone, Copy)]
+    pub(super) enum AttemptOutcome {
+        Ok,
+        Err,
+        Cancelled,
+    }
+
+    impl<T, E> From<&Result<T, E>> for AttemptOutcome {
+        fn from(value: &Result<T, E>) -> Self {
+            match value {
+                Ok(_) => AttemptOutcome::Ok,
+                Err(_) => AttemptOutcome::Err,
+            }
+        }
+    }
+
+    impl AttemptOutcome {
+        pub(super) fn as_str(&self) -> &'static str {
+            match self {
+                AttemptOutcome::Ok => "ok",
+                AttemptOutcome::Err => "err",
+                AttemptOutcome::Cancelled => "cancelled",
+            }
+        }
+    }
+
+    impl<C> PassFailCancelledRequestTyped<C> {
+        pub(super) fn get(&self, kind: RequestKind, outcome: AttemptOutcome) -> &C {
+            let target = match outcome {
+                AttemptOutcome::Ok => &self.success,
+                AttemptOutcome::Err => &self.fail,
+                AttemptOutcome::Cancelled => &self.cancelled,
+            };
+            target.get(kind)
+        }
+
+        fn build_with(mut f: impl FnMut(RequestKind, AttemptOutcome) -> C) -> Self {
+            let success = RequestTyped::build_with(|kind| f(kind, AttemptOutcome::Ok));
+            let fail = RequestTyped::build_with(|kind| f(kind, AttemptOutcome::Err));
+            let cancelled = RequestTyped::build_with(|kind| f(kind, AttemptOutcome::Cancelled));
+
+            PassFailCancelledRequestTyped {
+                success,
+                fail,
+                cancelled,
+            }
+        }
+    }
+
+    impl PassFailCancelledRequestTyped<Histogram> {
+        pub(super) fn observe_elapsed(
+            &self,
+            kind: RequestKind,
+            outcome: impl Into<AttemptOutcome>,
+            started_at: std::time::Instant,
+        ) {
+            self.get(kind, outcome.into())
+                .observe(started_at.elapsed().as_secs_f64())
+        }
+    }
+
+    pub(super) struct BucketMetrics {
+        /// Total requests attempted
+        // TODO: remove after next release and migrate dashboards to `sum by (result) (remote_storage_s3_requests_count)`
+        requests: RequestTyped<IntCounter>,
+        /// Subset of attempted requests failed
+        // TODO: remove after next release and migrate dashboards to `remote_storage_s3_requests_count{result="err"}`
+        failed: RequestTyped<IntCounter>,
+
+        pub(super) req_seconds: PassFailCancelledRequestTyped<Histogram>,
+        pub(super) wait_seconds: RequestTyped<Histogram>,
+        pub(super) cancelled_waits: RequestTyped<IntCounter>,
```

Comment:
Do we need to measure these? Maybe worth putting some rationale as a comment

Reply:
Oops, I meant to answer to this one with https://github.com/neondatabase/neon/pull/4892#discussion_r1284652279.

I think other cancellation reasons will be added later on with the lifecycle refactorings. Or maybe not.

Reply:
[9a4d8fd](https://github.com/neondatabase/neon/pull/4892/commits/9a4d8fd96b09df0b8b95bc449b3da74c5eb51b3c)

Reply:
Point is that we *hope* we don't have these, but we don't know. But I found that we do in fact do these right now (because of my mistake).

Reply:
Works for me, thanks!